File: problem_solving_with_DSA/stack/Python/implement-two-stack-in-single-array-method2.py
#!/usr/bin/env pyhton3
"""implement two stack in single array method2"""

size = 6
stack = [0 for _ in range(size)]

top1 = -1
top2 = size

    
def push1(val: int) -> None:
    """Push val into stack 1"""
    global top1 # update top1 from global scope
    
    if top1+1 == top2:
        return
    else:
        top1 += 1
        stack[top1] = val

def pop1() -> int:
    """Pop off the top data in the stack 1"""
    global top1
    
    if top1 == -1:
        return -1
    else:
        data = stack[top1]
        top1 -= 1
        return data
    

def push2(val: int) -> None:
    """Push val into stack 2"""
    global top2 # update top2 from global scope
    
    if top2-1 == top1:
        return
    else:
        top2 -= 1
        stack[top2] = val

def pop2() -> int:
    """Pop off the top data in the stack 2"""
    global top2
    
    if top2 == size:
        return -1
    else:
        data = stack[top2]
        top2 += 1
        return data



if __name__ == '__main__':
    push1(10)
    push1(20)
    push1(30)
    push1(40)
    push1(40)
    
    print(pop1())
    print(pop1())
    print(pop1())
    print(pop1())
    print(pop1())
    print(pop1())
    print(pop1())
    
    # push2(40)
    # push2(50)
    # # push2(60)
    # push2(70)
    
    # print(pop2())
    # print(pop2())
    # print(pop2())
    # print(pop2())
    # print(pop2())
    # print(pop2())

File: problem_solving_with_DSA/stack/C/implement-two-stack-in-single-array-method2.c
#include <stdio.h>

#define SIZE 6

int stack[SIZE];
int top1    = -1; // Top of stack 1
int top2 = SIZE; // Top of stack 2

/**
 * push1 - Push data into stack 1
 * 
 * @val: Value to push to stack 
 */
void push1(int val)
{
    if (top1 == top2-1) 
        return;
    stack[++top1] = val;
}

/**
 * pop1 - Pop data from the stack 1 
 * 
 * Return - Data popped of the stack
 */
int pop1()
{
    if (top1 == -1) 
        return -1;
    return stack[top1--];
}

/**
 * push2 - Push data into stack 2
 * 
 * @val: Value to push to stack 
 */
void push2(int val)
{
    if (top2-1 == top1) 
        return;
    stack[--top2] = val;
}



/**
 * pop2 - Pop data from the stack 2
 * 
 * Return - Data popped of the stack
 */
int pop2()
{
    if (top2 == SIZE) 
        return -1;
    return stack[top2++];
}

/**
 * printstack - Print data in stack 
 * 
 * @stack: Stack 
 * @size: Size of the stack 
 */
void printstack(int *stack, int size)
{
    printf("{");
    for (int i = 0; i < size; i++)
    {
        if (i < size-1)
            printf("%d, ", stack[i]);
        else
            printf("%d}\n", stack[i]);
    }
}

int main (void)
{   
    push1(10);
    push1(20);
    push1(30);
    
    printf("%d\n",pop1());
    printf("%d\n",pop1());
    printf("%d\n",pop1());
    printf("%d\n",pop1());
        
    push2(40);
    push2(50);
    push2(60);
    
    printf("%d\n",pop2());
    printf("%d\n",pop2());
    printf("%d\n",pop2());
    printf("%d\n",pop2());
    printf("%d\n",pop2());
    printf("%d\n",pop2());
    
    return (0);
}